### DISCUSSION TITLE
SPI Transfer Latency between the MCU and the Sx transceiver

### DISCUSSION THREAD

**Discussion by Zaher-Chaabene** [Lora-net/LoRaMac-node#1051] [Q&A] [answered]
Greetings,

I am working on a project where two transceivers (based on sx1262 and an ARM Cortex M4- Microcontroller) need to communicate with each other at the maximum transfer rate.

**Symptom :**
Assume at 125KHz of bandwidth, the achieved transfer-rate is _X_ Bytes/second, so theoretically the transfer-rate at 500KHz of bandwidth should be around 4*_X_ Bytes/second. 
But unfortunately, this is not the case as we are around only 2*_X_ Bytes/second.

**Root Cause :**
After some time analysis investigations, on the Receiving and Transmission windows on both end-points, I've noticed that there's a delay due to the SPI transfer on the reading/ writing functions from/ to the sx1262, and this delay is proportional to the transferred buffer size.  

![image](https://user-images.githubusercontent.com/78021430/108706939-a64dae00-750f-11eb-9829-94893129a654.png)

The problem is that the SPI transfer is performed byte by byte using the following for loop on the functions _SX126xWriteRegisters_, _SX126xReadRegisters_, _SX126xWriteBuffer_ and _SX126xReadBuffer_ of the sx1262mbxdas-board.c file :
`    for( uint16_t i = 0; i < size; i++ )
    {
        SpiInOut( &SX126x.Spi, buffer[i] );
    }`

**Workaround:**
I've had to re-write this function and replace the existing writing/ reading mechanism with the following HAL function : 
HAL_SPI_TransmitReceive();

This helped me accelerate the SPI transfer by approximately 80%, which is more suitable for my case. 

I would like to Know why using the for loop to read/ write byte by byte the buffer and is there anything wrong with the implementation of this workaround?  

Thanks!

**Comment by mluis1** [comment, marked as answer]
Nice to hear that you were able to solve your issue.

We have used the `for` loop because long time ago he had issues when using `HAL_SPI_TransmitReceive`.  
I can't remember anymore what were the issues.

In the meantime maybe ST has corrected the observed issues on the HAL or we have solved the issue without knowing it.

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.
